presto-hive/src/main/java/com/facebook/presto/hive/orc/OrcBatchPageSourceFactory.java

```diff
@@ -201,11 +201,10 @@ public static ConnectorPageSource createOrcPageSource(
         checkArgument(domainCompactionThreshold >= 1, "domainCompactionThreshold must be at least 1");
         checkRowIDPartitionComponent(columns, rowIDPartitionComponent);
 
-        OrcDataSource orcDataSource = getOrcDataSource(session, fileSplit, hdfsEnvironment, configuration, hiveFileContext, stats);
         Path path = new Path(fileSplit.getPath());
 
         OrcAggregatedMemoryContext systemMemoryUsage = new HiveOrcAggregatedMemoryContext();
-        try {
+        try (OrcDataSource orcDataSource = getOrcDataSource(session, fileSplit, hdfsEnvironment, configuration, hiveFileContext, stats)) {
```

Comment:
@sdruzkin - looking at your earlier comment "The reason why the page source factory is implemented as it's is that the life-cycle of the OrcDataSource is attached to the life-cycle of the created OrcPageSource. If you use try-catch-with-resource it will close the OrcDataSource right after creating the OrcPageSource and you won't be able to read anything from it" - I should **NOT** wrap the OrcDataSource orcDataSource in a try-catch-with resource here then.

Reply:
@sdruzkin- Thank you Sergii for your review and explanation of the close contract between the OrcDataSource, OrcReader, and OrcBatchPageSource and why a try-catch-with resource cannot be used here. This makes the refactor requested in this issue more trivial now.